src/shared/tpm2-util.c

```diff
@@ -307,55 +317,213 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+
+        /* Did Not find SRK, indicate this by returning 0 and ESYS_TR_NONE */
+        if (cap_data->data.handles.count == 0) {
+                *ret_primary = ESYS_TR_NONE;
+                return 0;
+        }
+
+        log_debug("Found SRK on TPM.");
+
+        /* convert the raw handle to an ESYS_TR */
+        TPM2_HANDLE handle = cap_data->data.handles.handle[0];
+        rc = sym_Esys_TR_FromTPMPublic(c,
+                        handle,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &primary);
+        if (rc != TSS2_RC_SUCCESS)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                        sym_Tss2_RC_Decode(rc));
+
+        /* Make sure the alg is what we were expecting */
+        _cleanup_(Esys_Freep) TPM2B_PUBLIC *out_public = NULL;
+        _cleanup_(Esys_Freep) TPM2B_NAME *out_name = NULL, *out_qname = NULL;
+        rc = sym_Esys_ReadPublic(
+                        c,
+                        primary,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &out_public,
+                        &out_name,
+                        &out_qname);
+        if (rc != TSS2_RC_SUCCESS)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                        sym_Tss2_RC_Decode(rc));
+
+        /* Cannot specify the alg and have a mismatch */
+        if (alg != 0 && alg != out_public->publicArea.type)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "SRK expected alg 0x%x does not match found alg 0x%x",
+                                        alg, out_public->publicArea.type);
+
+        *ret_primary = primary;
+
+        if (ret_alg)
+                 *ret_alg = out_public->publicArea.type;
+
+        return 0;
+}
+
+
+static int tpm2_make_transient_primary(
+                ESYS_CONTEXT *c,
+                ESYS_TR *ret_primary,
+                TPMI_ALG_PUBLIC alg,
+                bool use_srk_model,
+                TPMI_ALG_PUBLIC *ret_alg) {
+
+        static const TPM2B_SENSITIVE_CREATE primary_sensitive = {};
+        const TPM2B_PUBLIC *primary_template = NULL;
+
         static const TPML_PCR_SELECTION creation_pcr = {};
         ESYS_TR primary = ESYS_TR_NONE;
+        Tpm2SRKTemplateFlags base_flags = use_srk_model ? TPM2_SRK_TEMPLATE_NEW_STYLE : 0;
         TSS2_RC rc;
         usec_t ts;
 
-        log_debug("Creating primary key on TPM.");
+        /* Find existing SRK and use it if present */
+        if (use_srk_model) {
+                rc = tpm2_get_srk(c, ret_primary, alg, ret_alg);
```

Comment:
so as mentioned we have this rule that we want that for each function all return params are fully initiallized on success or not touched at all on failure.

now, here you pass ret_primary/ret_alg directly into tpm2_get_srk(), which will hence be modified on success. which is fine. but there are error paths further down here which might still fail, and if they do then you touched return params even on failure...

hence please use a local variable first, and only if the SRK already exists copy it from there into ret_primary/ret_alg.

(you already have `primary` and `alg` as local vars, so just use those)

Reply:
I also added asserts here and elsewhere now on things that callers cannot optionally pass for pointers.